packages/foundation/src/next/composed.tsx

```diff
@@ -42,8 +43,96 @@ export function composed<
   TComponentSlots = {},
   TStatics = {}
 >(
-  options: IComponentOptions<TComponentProps, TTokens, TStyleSet, TViewProps, TComponentSlots, TStatics> = {}
+  options: IComponentOptions<TComponentProps, TTokens, TStyleSet, TViewProps, TComponentSlots, TStatics>
+): React.FunctionComponent<TComponentProps> & TStatics;
+
+/**
+ * Recomposes a functional component based on the following set of options: styles, theme, view, and state.
+ * Imposes a separation of concern and centralizes styling processing to increase ease of use and robustness
+ * in how components use and apply styling and theming.
+ *
+ * Automatically merges and applies themes and styles with theme / styleprops having the highest priority.
+ * State component, if provided, is passed in props for processing. Props from state / user are automatically processed
+ * and styled before finally being passed to view.
+ *
+ * State components should contain all stateful behavior and should not generate any JSX, but rather simply call the view prop.
+ * Views should simply be stateless pure functions that receive all props needed for rendering their output.
+ * State component is optional. If state is not provided, created component is essentially a functional stateless component.
+ *
+ * @param baseComponent - base component to recompose
+ * @param options - component Component options. See IComponentOptions for more detail.
+ */
+export function composed<
+  TComponentProps extends ValidProps & ISlottableProps<TComponentSlots>,
+  TTokens,
+  TStyleSet extends IStyleSet<TStyleSet>,
+  TViewProps extends TComponentProps = TComponentProps,
+  TComponentSlots = {},
+  TStatics = {}
+>(
+  baseComponent: React.FunctionComponent,
+  options: IRecompositionComponentOptions<TComponentProps, TTokens, TStyleSet, TViewProps, TComponentSlots, TStatics>
+): React.FunctionComponent<TComponentProps> & TStatics;
+
+export function composed<
+  TComponentProps extends ValidProps & ISlottableProps<TComponentSlots>,
+  TTokens,
+  TStyleSet extends IStyleSet<TStyleSet>,
+  TViewProps extends TComponentProps = TComponentProps,
+  TComponentSlots = {},
+  TStatics = {}
+>(
+  baseComponentOrOptions:
+    | React.FunctionComponent
+    | IComponentOptions<TComponentProps, TTokens, TStyleSet, TViewProps, TComponentSlots, TStatics> = {},
+  recompositionOptions?: IRecompositionComponentOptions<TComponentProps, TTokens, TStyleSet, TViewProps, TComponentSlots, TStatics>
 ): React.FunctionComponent<TComponentProps> & TStatics {
+  const baseComponentOptions = (baseComponentOrOptions as ISlotCreator<
```

Comment:
oi; is this case needed? It really makes this harder to read.

Reply:
I think there are ways in which we can make this less cumbersome, I'll experiment.